pkg/jobmanager/jobmanager.go

```diff
@@ -45,3 +55,69 @@ func (jm *JobManager) LookupJob(id uuid.UUID) (*Job, error) {
 	}
 	return j, nil
 }
+
+func (jm *JobManager) StopJob(id uuid.UUID) error {
+	j, err := jm.LookupJob(id)
+	if err != nil {
+		return err
+	}
+	j.mu.Lock() // NOTE: We don't defer unlock as we want to wait after unlock unpon success.
+	if j.status != pb.JobStatus_JOB_STATUS_RUNNING {
+		j.mu.Unlock()
+		return nil
+	}
+	if j.cmd.Process != nil {
+		// Send the KILL to the process group, not just the process to avoid orphans.
+		// NOTE: This is POSIX compliant.
+		if err := syscall.Kill(-j.cmd.Process.Pid, syscall.SIGKILL); err != nil {
+			j.mu.Unlock()
+			if errors.Is(err, syscall.ESRCH) {
+				// If the process died as we were about to stop it, nothing to do.
+				// Don't set the status as stopped as it exited on it's own.
+				// This is an unavoidable "race" as we don't control the child process,
+				// it can die after the lock and before the kill. Nothing to worry about though.
+				return nil
+			}
+			return fmt.Errorf("process kill %d: %w", j.cmd.Process.Pid, err)
+		}
+	}
+	j.status = pb.JobStatus_JOB_STATUS_STOPPED
+	j.mu.Unlock()
+	<-j.waitChan
+	return nil
+}
+
+func (jm *JobManager) StreamLogs(ctx context.Context, id uuid.UUID) (io.Reader, error) {
+	j, err := jm.LookupJob(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a pipe for the caller to consume.
+	r, w := io.Pipe()
+
+	// Pause the broadcast, make a copy of the current output and subscribe the live logs.
+	//
+	// NOTE: Not ideal as it pauses the broadcast for every clients connected to that stream.
+	// In production, should consider a more advanced setup where we only pause the historical
+	// feed without pausing the stream of other clients.
```

Comment:
As suggested by @espadolini I updated this to use a strings.Builder, removing the copy.
This will still cause a 'lag' of few microseconds, probably less, for the context of the exercise I believe it would be acceptable.

If you disagree, I do have a different version of the broker that doesn't have this issue, however, it adds quite a lot of complexity.

Reply:
Docker has the same limitation, I figured if it was good enough for docker it was good eough for the exercise. https://github.com/moby/moby/blob/master/pkg/broadcaster/unbuffered.go#L23

Reply:
I missed some of the "live" discussion here (and in the other thread), but I think it's safe to say we would rather be resilient to a slow consumer, despite what Docker may do. This is one of the more interesting parts of the challenge for us.

Reply:
Changed the broadcaster to be non-blocking see (#8).